osquery/tables/system/windows/windows_search.cpp

```diff
@@ -0,0 +1,386 @@
+
+  CCommand<CDynamicAccessor, CRowset> cCommand;
+  hr = cCommand.Open(cSession, query.c_str());
+
+  if (FAILED(hr)) {
+    LOG(ERROR) << "error executing query";
+    return results;
+  }
+
+  for (hr = cCommand.MoveFirst(); S_OK == hr; hr = cCommand.MoveNext()) {
+    std::map<std::string, std::string> row;
+
+    for (DBORDINAL i = 1; i <= cCommand.GetColumnCount(); i++) {
+      auto columnName = cCommand.GetColumnName(i);
+      std::string columnNameStr = wstringToString(columnName);
+
+      // Add the column name and value to the row map.
+      row[columnNameStr] = ccomandColumnStringValue(cCommand, i);
+    }
+
+    results.push_back(row);
+  }
+
+  cCommand.Close();
+  return results;
+}
+
+LPCWSTR convertStringToLPCWSTR(std::string originalString) {
+    int stringLength = MultiByteToWideChar(CP_UTF8, 0, originalString.c_str(), -1, NULL, 0);
+    LPWSTR wideString = new WCHAR[stringLength];
+    MultiByteToWideChar(CP_UTF8, 0, originalString.c_str(), -1, wideString, stringLength);
+    return wideString;
+}
+
+std::string convertLPWSTRtoString(LPWSTR lpwstr) {
+    size_t len = wcslen(lpwstr);
+    int len_int = static_cast<int>(len);
+    int size_needed = WideCharToMultiByte(CP_UTF8, 0, lpwstr, len_int, NULL, 0, NULL, NULL);
+    std::string strTo(size_needed, 0);
+    WideCharToMultiByte(CP_UTF8, 0, lpwstr, len_int, &strTo[0], size_needed, NULL, NULL);
+    return strTo;
+}
+
+std::string generateSqlFromUserQuery(const std::string& userInput, std::string columns, std::string sort, LONG maxResults) {
+    HRESULT hr = NULL;
+
+    // Create ISearchManager instance
+    ISearchManager* pSearchManager = nullptr;
+    // Use library SearchSDK.lib for CLSID_CSearchManager.
+    hr = CoCreateInstance(CLSID_CSearchManager, NULL, CLSCTX_LOCAL_SERVER, IID_PPV_ARGS(&pSearchManager));
+    if (FAILED(hr)) {
+      LOG(ERROR) << "Failed to create ISearchManager instance";
+      return "";
+    }
+
+    // Create ISearchCatalogManager instance
+    ISearchCatalogManager* pSearchCatalogManager = nullptr;
+    // Call ISearchManager::GetCatalog for "SystemIndex" to access the catalog to the ISearchCatalogManager
+    hr = pSearchManager->GetCatalog(L"SystemIndex", &pSearchCatalogManager);
+    if (FAILED(hr)) {
+      LOG(ERROR) << "Failed to get catalog manager";
+      return "";
+    }
+
+    // Call ISearchCatalogManager::GetQueryHelper to get the ISearchQueryHelper interface
+    ISearchQueryHelper* pQueryHelper = nullptr;
+    hr = pSearchCatalogManager->GetQueryHelper(&pQueryHelper);
+    if (FAILED(hr)) {
+      LOG(ERROR) << "Failed to get query helper";
+      return "";
+    }
+
+    hr = pQueryHelper->put_QueryMaxResults(maxResults);
+    if (FAILED(hr)) {
+      LOG(ERROR) << "Failed to set max results";
+      return "";
+    }
+
+    if (!columns.empty()) {
+      hr = pQueryHelper->put_QuerySelectColumns(convertStringToLPCWSTR(columns));
+      if (FAILED(hr)) {
+        LOG(ERROR) << "Failed to set columns";
```

Comment:
I'm pretty sure these errors are exactly as you list them. So they should include something about the table name.

Reply:
done